### DISCUSSION TITLE
c#异步任务取消机制到底是如何运作的

### DISCUSSION THREAD

**Discussion by belZxx** [BYJRK/DotNet-Discussions#282] [C# 技术交流]
最近看了up主的异步任务相关视频,有几个问题想要咨询一下
问题1. WhenAll与WhenAny在任务取消时的运行机制,当await Task.WhenAll时,只有WhenAll中所有的任务完成时才会得到结果,那么如果此时提前取消其中一个任务,应该就会直接报错进入catch段落,可是结果好像和我预想的不一样,代码如下
```
var cts = new CancellationTokenSource();
var token = cts.Token;
cts.CancelAfter(1500);

var sw = Stopwatch.StartNew();

var CancelTask = Task.Run(async () => {
    await Task.Delay(2000);
    cts.Cancel();
});
try {
    await Task.WhenAll(Task.Delay(5000,  token), CancelTask);
    //await Task.WhenAny(Task.Delay(5000,  token), CancelTask);
}catch (Exception ex) {
    Console.WriteLine(ex.Message);
}
Console.WriteLine(sw.ElapsedMilliseconds);


//WhenAll output
A task was canceled.
2034

//WhenAny output
1555

```
当使用WhenAll时第一个Task.Delay的任务在1.5 s后取消,但还是等到了第二个CancelTask任务触发任务取消异常之后才进入了catch语句,也就是说这里的第一个任务取消并没有触发异常,而只是视作一个普通的任务完成
使用WhenAny得到的结论是相同的,第一个任务在1.5s 后取消,触发WhenAny,没触发异常,直接运行最后的输出语句
但是如果不使用WhenAny/WhenAll, 单独await Task.Delay(5000,  token)是能够正确触发异常的.所以是因为WhenAny/WhenAll只屏蔽了Task.Delay的任务取消的异常机制吗,因为之后CancelTask是能够正常取消任务并抛出异常的

问题2 自定义函数的任务取消机制无法触发,这里自定义了一个Foo函数,想要通过超时机制进行取消,但发现无法实现

```
void Foo() {
    Console.WriteLine("Foo Start");
    Console.ReadLine();
    Console.WriteLine("Foo End");
}

var cts = new CancellationTokenSource();
var token = cts.Token;
cts.CancelAfter(1500);

var sw = Stopwatch.StartNew();

try {
   await Task.Factory.StartNew(Foo, token);
}catch (Exception ex) {
    Console.WriteLine(ex.Message);
}

Console.WriteLine(sw.ElapsedMilliseconds);

//output
Foo Start

Foo End
11208 
```

可以发现经过1.5s任务没有取消,直到按下回车键才会结束,这又是为什么呢

**Comment by hzlcs** [comment]
因为await WhenAll返回的是void，而WhenAny返回的是一个Task，所以其实WhenAny的异常在它返回的task里，也就是 
var t = await Task.WhenAny(); no exception
await t;  A task was canceled.
这么设计应该是因为WhenAny时你可能需要知道首先完成的是哪个任务，然后再去处理结果或异常

**Reply by hzlcs** [reply]
https://stackoverflow.com/questions/31544684/why-does-the-task-whenany-not-throw-an-expected-timeoutexception

**Reply by belZxx** [reply]
原来如此,我明白了,感谢回复

**Comment by hzlcs** [comment]
第二个问题是在自定义函数里要有自己的超时机制，并将外部的token传入才能起效
而Task.Factory.StartNew这种只能对加入Factory但还未开始的task起效，一旦开始这个token就是无效的
Task.Run还有可能起效，而Task.Factory.StartNew直接就是一个开始的task，除非这个token已经是取消的状态才会起效
可以试一下把cts.CancelAfter(1500);改成cts.Cancel()或cts.CancelAfter(1);就会抛异常了;

**Reply by belZxx** [reply]
感谢回复,我尝试了一下Task.Factory.StartNew搭配cts.Cancel()确实可以抛异常,不过cts.CancelAfter(1)没有抛异常,但是这样不能起到想要的超时目的.
Task.Run的结果和Task.Factory.StartNew一样,也不会取消.不过尝试了一下使用Task.Factory.StartNew搭配Wait是可以实现的
Task.Factory.StartNew(Foo).Wait(3000);

**Reply by hzlcs** [reply]
自定义函数想要真正去取消可以像这样
```c#
async Task Foo(CancellationToken token)
{
    Console.WriteLine("Foo Start");
    TaskCompletionSource s = new();
    IAsyncDisposable disposable = token.Register(s.SetCanceled);
    Task t = Task.Run(() =>
    {
        Console.ReadLine();
        try
        {
            token.ThrowIfCancellationRequested();
        }
        catch
        {
            //rollback
        }
        Console.WriteLine("Foo End");
    }, token);

    Task res = await Task.WhenAny(t, s.Task);
    await disposable.DisposeAsync();
    await res;
}

var cts = new CancellationTokenSource();
var token = cts.Token;
cts.CancelAfter(1500);

var sw = Stopwatch.StartNew();

try
{
    await Foo(token);
}
catch (Exception ex)
{
    Console.WriteLine(ex.Message);
}

Console.WriteLine(sw.ElapsedMilliseconds);
```

**Reply by belZxx** [reply]
学到了,感谢分享
